pkgs/by-name/pr/protonmail-bridge-gui/package.nix

```diff
@@ -0,0 +1,105 @@
+{ lib
+, stdenv
+, pkg-config
+, libsecret
+, cmake
+, ninja
+, qt6
+, grpc
+, protobuf
+, zlib
+, gtest
+, sentry-native
+, protonmail-bridge
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "protonmail-bridge-gui";
+
+  inherit (protonmail-bridge) version src;
+
+  patches = [
+    ./use-nix-googletest.patch
+    ./designator-order-fix.patch
+  ];
```

Comment:
Each patch in nixpkgs applied to the upstream source should be documented. Out-of-tree patches, fetched using `fetchpatch`, are preferred.

In each patch comment, please explain the purpose of the patch and link to the relevant upstream issue if possible. If the patch has been merged upstream but is not yet part of the released version, please note the version number or date in the comment such that a future maintainer updating the nix expression will know whether the patch has been incorporated upstream and can thus be removed from nixpkgs.

Furthermore, please use a _stable_ URL for the patch. Rather than, for example, linking to a GitHub pull request of the form `https://github.com/owner/repo/pull/pr_number.patch`, which would change every time a commit is added or the PR is force-pushed, link to a specific commit patch in the form `https://github.com/owner/repo/commit/sha.patch`.

Here are two good examples of patch comments:

```nix
mkDerivation {
  …

  patches = [
    # Ensure RStudio compiles against R 4.0.0.
    # Should be removed next 1.2.X RStudio update or possibly 1.3.X.
    (fetchpatch {
      url = "https://github.com/rstudio/rstudio/commit/3fb2397c2f208bb8ace0bbaf269481ccb96b5b20.patch";
      sha256 = "0qpgjy6aash0fc0xbns42cwpj3nsw49nkbzwyq8az01xwg81g0f3";
    })
  ];
}
```

```nix
mkDerivation {
  …

  patches = [
    # Allow building with bison 3.7
    # PR at https://github.com/GoldenCheetah/GoldenCheetah/pull/3590
    (fetchpatch {
      url = "https://github.com/GoldenCheetah/GoldenCheetah/commit/e1f42f8b3340eb4695ad73be764332e75b7bce90.patch";
      sha256 = "1h0y9vfji5jngqcpzxna5nnawxs77i1lrj44w8a72j0ah0sznivb";
    })
  ];
}
```

Reply:
The patches are Nix specific workarounds that I won't be requesting a merge for upstream, but I'll add explanatory comments. There are open PRs [re: offline builds](https://github.com/ProtonMail/proton-bridge/pull/441), and even re: the [antiquated ICU version](https://github.com/ProtonMail/proton-bridge/pull/347) but I think my patches are simpler and more specific to Nix.

Reply:
Fine, just that comment is useful for the next person who's going to eventually review this file !

Reply:
By the way, I take it you used https://github.com/jtojnar/nixpkgs-hammering?

It looks handy, I'd love to use it, but I'm probably missing a trick, can you share your commands on how to use it on this derivation, I get

```
[daniel@laptop:~/Source/nixpkgs]$ nix run github:jtojnar/nixpkgs-hammering protonmail-bridge-gui
warning: unknown experimental feature 'configurable-impure-env'
When evaluating attribute ‘protonmail-bridge-gui’:
warning: EvalError
Cannot evaluate attribute ‘protonmail-bridge-gui’ in ‘/nix/store/kpp5lpaykmyzfw5ml14ikjmdv3hipbpw-nixpkgs’.
notice: no-build-output
‘protonmail-bridge-gui’ has not yet been built. Checks that rely on the presence of build logs are skipped.
```

Reply:
No, I don't use it, just using the helpful review messages. Maybe you can ask @jtojnar if you need help with it.

Reply:
The command is okay, some of the checks are a bit fragile, fixed in https://github.com/jtojnar/nixpkgs-hammering/commit/1e799b6cdba138c0d8497268865e33b179376719

Reply:
Thanks @jtojnar I think, as the message indicated, I just needed to build it first

```
[daniel@laptop:~/Source/nixpkgs]$ nix run github:jtojnar/nixpkgs-hammering protonmail-bridge-gui
warning: unknown experimental feature 'configurable-impure-env'
warning: unknown experimental feature 'configurable-impure-env'
warning: unknown experimental feature 'configurable-impure-env'
When evaluating attribute ‘protonmail-bridge-gui’:
warning: attribute-ordering
The lists of dependencies, including the attribute “nativeBuildInputs” should preferably come before build system configuration flags’ “cmakeFlags” attribute in the expression.

Near /home/daniel/Source/nixpkgs/pkgs/by-name/pr/protonmail-bridge-gui/package.nix:21:3:
   |
21 |   cmakeFlags = [
   |   ^
Near /home/daniel/Source/nixpkgs/pkgs/by-name/pr/protonmail-bridge-gui/package.nix:31:3:
   |
31 |   nativeBuildInputs = [
   |   ^
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-ordering.md
warning: attribute-ordering
The attribute “patches” should preferably come before lists of dependencies’ “buildInputs” attribute in the expression.

Near /home/daniel/Source/nixpkgs/pkgs/by-name/pr/protonmail-bridge-gui/package.nix:42:3:
   |
42 |   buildInputs = [
   |   ^
Near /home/daniel/Source/nixpkgs/pkgs/by-name/pr/protonmail-bridge-gui/package.nix:53:3:
   |
53 |   patches = [
   |   ^
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-ordering.md
warning: explicit-phases
It is a good idea to avoid overriding `installPhase` when possible.

Near /home/daniel/Source/nixpkgs/pkgs/by-name/pr/protonmail-bridge-gui/package.nix:69:3:
   |
69 |   installPhase = ''
   |   ^
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/explicit-phases.md
```

Reply:
Hi, shimming in this thread but to me it look like `designator-order-fix.patch` doesn't seems that much nix specifics to me but rather that nix is using a more up to date/slightly different compiler than what is used in proton-bridge's build process/host.

Reply:
This is the version of GCC they compiled the binaries in their Debian release with:
```
[nix-shell:~/Downloads/bridge/usr/lib/protonmail/bridge]$ readelf -p .comment bridge-gui

String dump of section '.comment':
  [     0]  GCC: (Debian 10.2.1-6) 10.2.1 20210110
```

So I tried `gcc10Stdenv` (gives GCC 10.5.0) in place of `stdenv` and even `gcc9Stdenv` in this derivation, still get this error without the patch:
```
protonmail-bridge-gui> FAILED: bridge-gui-tester/CMakeFiles/bridge-gui-tester.dir/GRPCService.cpp.o
protonmail-bridge-gui> /nix/store/qhpw32pz39y6i30b3vrbw5fw6zv5549f-gcc-wrapper-13.2.0/bin/g++ -DBRIDGE_APP_VERSION=\"3.9.1\" -DPROTOBUF_USE_DLLS -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_QMLINTEGRATION_LIB -DQT_QML_LIB -DQT_WIDGETS_LIB -I/build/source/internal/frontend/bridge-gui/build/bridge-gui-tester/bridge-gui-tester_autogen/include -I/build/source/internal/frontend/bridge-gui/bridge-gui-tester -I/build/source/internal/frontend/bridge-gui/bridgepp -isystem /nix/store/vcg2izx98281pckvs280h5hhhy61kf6n-qtbase-6.6.2/include/QtCore -isystem /nix/store/vcg2izx98281pckvs280h5hhhy61kf6n-qtbase-6.6.2/mkspecs/linux-g++ -isystem /nix/store/vcg2izx98281pckvs280h5hhhy61kf6n-qtbase-6.6.2/include/QtGui -isystem /nix/store/vcg2izx98281pckvs280h5hhhy61kf6n-qtbase-6.6.2/include/QtWidgets -isystem /nix/store/pjrzj2pkzgaghw7id8wi54i41gazh1gi-qtdeclarative-6.6.2/include/QtQml -isystem /nix/store/pjrzj2pkzgaghw7id8wi54i41gazh1gi-qtdeclarative-6.6.2/include/QtQmlIntegration -isystem /nix/store/vcg2izx98281pckvs280h5hhhy61kf6n-qtbase-6.6.2/include/QtNetwork -O3 -DNDEBUG -std=gnu++20 -fPIC -Winvalid-pch -include /build/source/internal/frontend/bridge-gui/build/bridge-gui-tester/CMakeFiles/bridge-gui-tester.dir/cmake_pch.hxx -MD -MT bridge-gui-tester/CMakeFiles/bridge-gui-tester.dir/GRPCService.cpp.o -MF bridge-gui-tester/CMakeFiles/bridge-gui-tester.dir/GRPCService.cpp.o.d -o bridge-gui-tester/CMakeFiles/bridge-gui-tester.dir/GRPCService.cpp.o -c /build/source/internal/frontend/bridge-gui/bridge-gui-tester/GRPCService.cpp
protonmail-bridge-gui> /build/source/internal/frontend/bridge-gui/bridge-gui-tester/GRPCService.cpp: In member function 'virtual grpc::Status GRPCService::RequestKnowledgeBaseSuggestions(grpc::ServerContext*, const google::protobuf::StringValue*, google::protobuf::Empty*)':
protonmail-bridge-gui> /build/source/internal/frontend/bridge-gui/bridge-gui-tester/GRPCService.cpp:368:30: error: designator order for field 'bridgepp::KnowledgeBaseSuggestion::url' does not match declaration order in 'bridgepp::KnowledgeBaseSuggestion'
protonmail-bridge-gui>   368 |         suggestions.push_back(        {
protonmail-bridge-gui>       |         ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
protonmail-bridge-gui>   369 |             .title = QString("Suggested link %1").arg(i),
protonmail-bridge-gui>       |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
protonmail-bridge-gui>   370 |             .url = QString("https://proton.me/support/bridge#%1").arg(i),
protonmail-bridge-gui>       |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
protonmail-bridge-gui>   371 |         });
protonmail-bridge-gui>       |         ~~
```

I couldn't even get it to build on a Debian VM because of https://github.com/ProtonMail/proton-bridge/issues/346. They don't provide their build scripts (mentioned [here](https://github.com/ProtonMail/proton-bridge/issues/462#issuecomment-1963568277)), nor a CI/CD pipeline log, nor a Dockerfile. Without their build scripts and/or Dockerfile it's difficult to say whether they get it to build with some compiler flag or if like you say, run into the same problem. Without even being able to reproduce their build environment I decided against opening a PR upstream at this stage. I'll update the comment next to the patch on the derivation, and maybe open an issue with them.

Reply:
@MrFreezeex I had another look, and I'm fairly certain it's an upstream bug in the C++ of the `bridge-gui-tester` component of the GUI, a development tool that can be safely excluded from the build. I just removed `add_subdirectory(bridge-gui-tester)` from the main `CMakeLists.txt`, and it results in a smaller, faster build too! See https://github.com/NixOS/nixpkgs/blob/df4ffc959dda485ed0f81d2fdd1c8525f6034b19/pkgs/by-name/pr/protonmail-bridge-gui/package.nix#L62